packages/tideline/js/tidelinedata.js

```diff
@@ -14,1010 +14,1090 @@
-    // the fill should extend past the *end* of a segment (i.e. of basal data)
-    // if that's the last datum in the data
-    var lastTimestamp = lastDatum.normalEnd || lastDatum.normalTime;
-    var first = new Date(this.data[0].normalTime), last = new Date(lastTimestamp);
-    // make sure we encapsulate the domain completely
-    if (last - first < MS_IN_DAY) {
-      first = d3.time.hour.utc.offset(first, -12);
-      last = d3.time.hour.utc.offset(last, 12);
+    if (!_.isString(pa.inputTime) || _.isEmpty(pa.inputTime)) {
+      pa.inputTime = pa.normalTime;
     }
-    else {
-      first = d3.time.hour.utc.offset(first, -6);
-      last = d3.time.hour.utc.offset(last, 6);
-    }
-    this.grouped.fill = fillDataFromInterval(first, last);
-    endTimer('generateFillData');
-    return this;
+  });
+
+  // get all PAs grouped by eventID
+  const physicalActivities = _.groupBy(physicalActivity, "eventId");
+  // For each eventID sort by inputTime
+  _.forEach(physicalActivities, (value, key) => {
+    physicalActivities[key] = _.orderBy(value, ["inputTime"], ["desc"]);
+  });
+  // For each eventID take the most recent item
+  this.physicalActivities = _.map(physicalActivities, _.head);
+};
+
+TidelineData.prototype.setDiabetesData = function setDiabetesData() {
+  const diabetesDataTypes = this.opts.diabetesDataTypes ?? DIABETES_DATA_TYPES;
+  this.diabetesData = this.data.filter((d) => diabetesDataTypes.indexOf(d.type) > -1);
+  this.diabetesData = _.sortBy(this.diabetesData, "normalTime");
+};
+
+/**
+ *
+ * @param {Datum[]} data The data
+ * @param {object} filter The filter
+ * @param {string[]} order field to order the result
+ */
+TidelineData.prototype.setEvents = function setEvents(filter = {}, order = ["inputTime"]) {
+  const sourceEvents = _.groupBy(_.filter(this.data, filter), "eventId");
+  const events = {};
+  _.forEach(sourceEvents, function (value, key) {
+    events[key] = _.orderBy(value, order, ["desc"]);
+  });
+  // eslint-disable-next-line lodash/prop-shorthand
+  return _.map(events, (v) => v[0]);
+};
+
+TidelineData.prototype.getLastestManufacturer = function getLastestManufacturer() {
+  const defaultPumpManufacturer = {
+    payload: { pump: { manufacturer: "default" } },
   };
 
-  // two-week view requires background fill rectangles from midnight to midnight
-  // for each day from the first through last days where smbg exists at all
-  // and for at least 14 days
-  this.adjustFillsForTwoWeekView = function() {
-    startTimer('adjustFillsForTwoWeekView');
-    var fillData = this.grouped.fill;
-    var endpoints = getTwoWeekFillEndpoints();
-    this.twoWeekData = this.grouped.smbg || [];
-    var twoWeekFills = fillDataFromInterval(new Date(endpoints[0]), new Date(endpoints[1]), false);
-    this.twoWeekData = _.sortBy(this.twoWeekData.concat(twoWeekFills), function(d) {
-      return d.normalTime;
+  if (!Array.isArray(this.grouped.pumpSettings)) {
+    return defaultPumpManufacturer;
+  }
```

Comment:
@yvon-dblg 
This has to be fixed in later version. 
If we have no settings (I know that should not happen) the default pump manufacturer is not added to the deviceEvents. At the end the reservoir  tooltip is broken.